### DISCUSSION TITLE
creating-a-source-generator-part-9-avoiding-performance-pitfalls-in-incremental-generators/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#405] [Announcements]
# creating-a-source-generator-part-9-avoiding-performance-pitfalls-in-incremental-generators/

In this post I describe some important things to think about when designing your incremental source generator, particularly the pitfalls to watch out for

https://andrewlock.net/creating-a-source-generator-part-9-avoiding-performance-pitfalls-in-incremental-generators/

**Comment by bakgaard** [comment]
These are good points!

Do you have snippets to test if the generated code is faster than the original? So one can verify (and justify to a colleague) that it is worth spending time on?

**Reply by MichalBrylka** [reply]
I am always providing proof I.e. 
https://github.com/nemesissoft/Nemesis.TextParsers/blob/08fc4d8e15273826cdfe2e8840c208a01563cf3a/Benchmarks/EnumBenchmarks/EnumParserBench_CodeGen.cs#L17C14-L17C14

Otherwise no one would support using such not-so-friendly techniques 😎

**Comment by psulek** [comment]
Really helpful series! Now I can start writing my first source code generator 👍

**Comment by MichalBrylka** [comment]
Nice summary. How in your opinion should we handle Diagnostics in Incremental Generators?
My approach was to use `Result<InputForGenerator, Microsoft.CodeAnalysis.Diagnostic>` (discriminate union) as a return type for "transform" delegate.  My `InputForGenerator` is record struct and Diagnostic seem to be a good value-semantics citizen

**Reply by andrewlock** [reply]
That's a good question - I'm going to add another section to cover diagnostics, but unfortunately you shouldn't use Diagnostic directly, instead you need to extract the data you need. 

The DU bit is good though, I use that too 😊

**Reply by MichalBrylka** [reply]
You mean like instead of storing Diagnostic I should return DU like Result<Input, RecordContainingRuleAndLocationAndOtherStuff>

**Comment by adampaquette** [comment]
I would love to know what you think about my project https://github.com/adampaquette/Typely. It has similarities with StronglyTypedId. It generates domain primitives bounded to validations and remove a lot of boiler plates from Api validation and conversions to EF Core configuration. The primary goal was to make value objects as performant as primitives like int, bool and avoid as much memory allocation as possible. The second goal was to make domain primitives as accessible as their primitives counter parts and avoid situations like "we will not implement domain primitives because there is too much work and overhead". The generator has been reviewed by some members of Microsoft and should be performant. A sample application is available here : https://github.com/adampaquette/Typely/tree/main/sample.

**Comment by StefanOssendorf** [comment]
Great guide! I'm writing my first source gen and this series is tremendously helpful. Thank you for that :-).

I'm struggling with the `ForAttributeWithMetadataName` approach. It's working for my marker Attribute but when using an attribute from a 3rd Party Library it does not get called even though my manual approach (predicate which checks if it's the expected attribute) works. Any ideas/hints for that?

**Comment by drdamour** [comment]
any chance on a post dedicated to diagnostics?  i've never understood if i should error/warn when the types annotated by my generators driving attribute are not partial.  seems like best practices around this are lacking in the community

**Comment by Seekatar** [comment]
This is a great series. I created a [source generator](https://github.com/Seekatar/OptionToStringGenerator) from it and am looking forward to implementing these changes. 

You mention using 4.4.0+ of Microsoft.CodeAnalysis.CSharp. I originally used 4.6.0 and had problems with users that have older .NET 6 SDKs installed getting an error like the following and had to revert to 4.3.0 to fix it. 

```
CSC : warning CS8032: An instance of analyzer Seekatar.OptionToStringGenerator.OptionPropertyToStringGenerator cannot be created from ... Could not load file or assembly 'Microsoft.CodeAnalysis, Version=4.3.0.0,
```

Will using 4.4.0 cause similar problems, and do you have advice for fixing it? Googling yields various hacks but I found nothing definitive.

**Comment by DemoBytom** [comment]
How, in your opinion, should we handle situations where code generator output would depend on code in two different assemblies.
Let's say I have a Core library with entity models, and a Repository library that references the Core, and contains a repository implementation.
Now let's say I want a code generator that generates some code in Repository for every entity model in the Core library. Reflection is the only way I found so far, but I guess reflection should rather be avoided? Is there a way to do it in some other way?

**Reply by damienhoneyford** [reply]
You might be able to employ a similar method that [Microsoft use](https://github.com/dotnet/runtime/blob/v8.0.4/src/libraries/System.Text.Json/gen/Helpers/KnownTypeSymbols.cs) in the System.Text.Json Source Generator (assuming I understand what it's doing here, and that it would work for code that's not part of the BCL)

**Comment by damienhoneyford** [comment]
Your point on the use of the CompilationProvider is interesting, given [Microsoft's use of it](https://github.com/dotnet/runtime/blob/2d7eea252964e69be94cb9c847b371b23e4dd470/src/libraries/System.Text.Json/gen/JsonSourceGenerator.Roslyn4.0.cs#L31) in the System.Text.Json Source Generator.

The [KnownTypeSymbols](https://github.com/dotnet/runtime/blob/v8.0.4/src/libraries/System.Text.Json/gen/Helpers/KnownTypeSymbols.cs) class carries a copy of the Compilation around with it and is used by [Parser](https://github.com/dotnet/runtime/blob/v8.0.4/src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs) at [this point](https://github.com/dotnet/runtime/blob/2d7eea252964e69be94cb9c847b371b23e4dd470/src/libraries/System.Text.Json/gen/JsonSourceGenerator.Roslyn4.0.cs#L45).

I'd be interested to know what your view of their Source Generator is -- would it suffer any 'performance pitfalls'?

**Reply by andrewlock** [reply]
> would it suffer any 'performance pitfalls

As far as I can tell, yes. 

They have this:

```csharp
IncrementalValueProvider<KnownTypeSymbols> knownTypeSymbols = context.CompilationProvider
                .Select((compilation, _) => new KnownTypeSymbols(compilation));
```

Which will output every single time something about the `Compilation` changes. And as `KnownTypeSymbols` is a class, it will be a _different_ instance every time, so as I understand it, will be considered a _Different_ object every time.

They then use `ForAttributeWithMetadataName` in the main path, which is good:

```csharp
 context.SyntaxProvider.ForAttributeWithMetadataName(...)
```

But then they use it with `Combine(knownTypeSymbols)` which will [invoke the output](https://github.com/dotnet/roslyn/blob/main/docs/features/incremental-generators.md#combine) every time the `Compilation` changes

So unless I'm missing something, yeah, it's doing a _lot_ of work 🤷‍♂️

**Reply by damienhoneyford** [reply]
I think the issue of the `Compilation` class being recreated all the time will likely also extend to the `CompilationOptions` class too as it has an internal field named `CurrentLocalTime` which is included as part of its equality check logic -- likely making [DiagnosticDescriptor.GetEffectiveSeverity(CompilationOptions)](https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.diagnosticdescriptor.geteffectiveseverity?view=roslyn-dotnet-4.7.0) almost too costly to consider using.

I also had a quick look in to whether `Location` would be cacheable, and I think it is, at least when coupled with [this helper method](https://github.com/dotnet/runtime/blob/f0991722db63e0b4962439469efb4ea1b72923f3/src/libraries/Common/src/SourceGenerators/DiagnosticInfo.cs#L33) I found in Microsoft's System.Text.Json Source Generator:

```csharp
// Creates a copy of the Location instance that does not capture a reference to Compilation.
static Location GetTrimmedLocation(Location location)
    => Location.Create(location.SourceTree?.FilePath ?? "", location.SourceSpan, location.GetLineSpan().Span);
```

Overall, it's a real shame Microsoft don't elaborate on the best practices for Source Generators -- their current documentation is definitely lacking, but I've found your series a real help in developing my own Source Generator, so thank you.

**Comment by m31coding** [comment]
Hi Andrew, 

Thank you for this! An excellent write-up, as always. I am considering updating the M31.FluentAPI to make use of the ForAttributeWithMetadataName method. However, I am confused about the dependencies. This feature was introduced in Microsoft.CodeAnalysis.CSharp 4.3.1, which, according to the [Roslyn docs](https://github.com/dotnet/roslyn/blob/main/docs/wiki/NuGet-packages.md), is part of .NET 6. On the other hand, I've  read in several sources that .NET 7 is required. Gaining the performance of the new method but staying compatible with .NET 6 seems to be a good choice for now, if possible. 

Best regards,
Kevin

**Reply by damienhoneyford** [reply]
`ForAttributeWithMetadataName` needs the .NET 7 SDK installed, but the project your Source Generator is used in can still target .NET 6 -- this is exactly my use case, and made me wonder why the `ForAttributeWithMetadataName` method was showing up as it didn't when I first started writing the Source Generator, which is .NET 6 and the answer was as simple as 'because I didn't have the .NET 7 SDK installed at the time'.

Keep in mind though that .NET 6 is only supported until the end of the year and .NET 7 support runs out in mess than a month.

**Reply by m31coding** [reply]
Thank you for your insights, this helps me to decide how to move forward.

**Reply by m31coding** [reply]
Does anyone know how to learn about these SDK requirements without trial and error? [This page](https://github.com/dotnet/roslyn/blob/main/docs/wiki/NuGet-packages.md) states that the Roslyn packages of version 4.3.1 require Visual Studio 2022 version 17.3 and .NET 6. According to [this page](https://learn.microsoft.com/en-us/dotnet/core/porting/versioning-sdk-msbuild-vs#lifecycle), VS 17.3 comes with SDK version 6.0.4xx. Unfortunately I was unable to find information stating that the .NET 7 SDK is required.

**Comment by srogovtsev** [comment]
> Don't use *Syntax or ISymbol instances in your pipeline
> Be careful using CompilationProvider

If we take a look at Microsoft's own `LoggerMessageGenerator`, they do both: https://github.com/dotnet/runtime/blob/7ac5ef959579da0202e6ad618924dfc8fe129e89/src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Roslyn4.0.cs#L34-L35

```
IncrementalValueProvider<(Compilation, ImmutableArray<ClassDeclarationSyntax>)> compilationAndClasses = context.CompilationProvider.Combine(classDeclarations.Collect());
```

Does this mean I'm completely misunderstanding your recommendations, or they have changed something since this was written, or they just don't care about the performance in this run path?

I'm trying to write my own generator for a non-critical path and currently trying to understand which approach would be better.

**Reply by andrewlock** [reply]
> they just don't care about the performance in this run path

IMO, how to write a _performant_ incremental source generator was not well understood by anyone outside of the Roslyn team when they were first introduced. Unfortunately that means even source generators built into the runtime have poor performance characteristics 😬 

- https://github.com/dotnet/runtime/issues/93309
- https://github.com/dotnet/runtime/issues/93313
- https://github.com/dotnet/runtime/issues/92914

> I'm trying to write my own generator for a non-critical path and currently trying to understand which approach would be better.

Definitely stick to my recommendations here if you can 🙂 Nothing's changed in that regard as far as I know.

**Reply by srogovtsev** [reply]
Okay, thanks for the clarification!

PS By the way, as I have an occasion to do so, thank you for your "ASP.NET Core in Action", which I enjoyed greatly.

**Reply by andrewlock** [reply]
Thank you, and you're welcome!

**Comment by mungojam** [comment]
Thanks for this great series, ForAttributeWithMetadataName didn't seem to work as soon as I added some constructor parameters to the attribute. I wonder if I need to include those parameters in the attribute name I pass to it somehow

**Reply by andrewlock** [reply]
Hmmm, no, that shouldn't be required I think 🤔

**Comment by AndersKehlet** [comment]
Diagnostics in generators was discussed on the C# discord server and the "official" word is "don't". 

From Cyrus:

>this is not to make some flow chart that has a narrow path by which someone can go: a-ha!  i can put diagnostics into my generator!
It's to get people off of doing this entirely.  note: we're having the discussion right now about just dropping support for diagnostics in generators entirely.
(and i'm a big proponent of that, because continuing to have this sort of back and forth on if there are scenarios where it is ok to do this, is just not a good use of time imo)

**Reply by andrewlock** [reply]
Interesting - I don't suppose you have any more details (on GitHub ideally)? I've not heard this discussed anywhere before!

**Reply by Serdan** [reply]
https://discord.com/channels/143867839282020352/598678594750775301

#roslyn channel, february 23rd

There are several Roslyn people in there, so I think it's easiest to just get an answer straight from the horse's mouth, as it were.

**Reply by andrewlock** [reply]
Urgh, you're going to make me join discord?! 😥😉😛

**Reply by AndersKehlet** [reply]
I don't think anyone would mind if you created a new discussion on the roslyn repo and tagged @CyrusNajmabadi 😄

**Reply by CyrusNajmabadi** [reply]
basically, there are no efficient ways of doing diagnostics with incremental generation.  so the strong preference is that you do diagnostics with analzyers so that generators can be entirely incremental and almost never need to recompute/regen anything.  That's basically it :)

**Reply by MichalBrylka** [reply]
@CyrusNajmabadi would you be able to point us to a model implementation where analyzers work in tandem with incremental generators ?

**Reply by AndersKehlet** [reply]
@MichalBrylka, there's no magic too it. You just bundle an analyzer with the generator in the nuget package.

I'm working on a template with all the best practice stuff I've learned over the past few months.

https://github.com/Serdan/SourceGeneratorTemplate

Feel free to use it however you want.

**Comment by LordZoltan** [comment]
Just a quick thing on the stress over the GetHashCode() implementation: you don't need it for Roslyn's pipeline caching; you only need it if you intend to build hashsets/dictionaries of your model types.

**Comment by sebastianstudniczek** [comment]
Great post! 
What are your toughts on getting objects instance values inside source generator? I am writing source gen to simplify dependency injection and try to use classes property value as key for di registration.

**Reply by bkoelman** [reply]
Thanks for the great tips. Here's another one.

To correctly and efficiently ensure unique file names (which is a requirement for generators), see https://github.com/dotnet/roslyn/discussions/60272#discussioncomment-6053422. However, the referenced code contains a bug that can be fixed with a 1-line change; see https://github.com/dotnet/roslyn/issues/80813.

**Comment by robverp78** [comment]
Awesome series — I wish I'd read it before I went experimenting myself :D

I saw there was some discussion about diagnostics and wondered if you're still planning to write that part. It's something I kept running into: needing to return a diagnostic during transformation, and then discovering only later that one was needed — by which point you have to refactor both the transform result and the output to carry it. And you still need to handle cases where there's a result but with a warning, or no result but an error, or some other variation.

I ended up building a small library for it: https://www.nuget.org/packages/Subro.Generators.TransformResult/

Basically, any transform function returns a lightweight TransformResult<T> struct. Instead of a nullable, you return default for empty — but T and diagnostics are implicitly converted, so you can return any combination cleanly. Helper functions then split results and diagnostics into their own paths.

Helper functions then split results and diagnostics into their own paths — and crucially, the diagnostics are registered automatically, so you don't need a separate RegisterSourceOutput call for them. The code generation just receives the valid results and the rest is handled.

Funnily enough, I only noticed the diagnostics section of your series just now, and some of the naming overlaps :D I suppose suffixing things with Info is a pretty natural instinct.
